db/compaction/compaction_job.cc

```diff
@@ -1773,198 +1294,47 @@ Status CompactionJob::FinishCompactionOutputFile(
-        // min key in the following table or subcompaction) in order for files
-        // to appear key-space partitioned.
-        //
-        // Choose highest seqnum so this file's largest internal key comes
-        // before the next file's/subcompaction's smallest. The fake seqnum is
-        // OK because the read path's file-picking code only considers the user
-        // key portion.
-        //
-        // Note Seek() also creates InternalKey with (user_key,
-        // kMaxSequenceNumber), but with kTypeDeletion (0x7) instead of
-        // kTypeRangeDeletion (0xF), so the range tombstone comes before the
-        // Seek() key in InternalKey's ordering. So Seek() will look in the
-        // next file for the user key.
-        largest_candidate =
-            InternalKey(*upper_bound, kMaxSequenceNumber, kTypeRangeDeletion);
-      }
-#ifndef NDEBUG
-      SequenceNumber smallest_ikey_seqnum = kMaxSequenceNumber;
-      if (meta->smallest.size() > 0) {
-        smallest_ikey_seqnum = GetInternalKeySeqno(meta->smallest.Encode());
-      }
-#endif
-      meta->UpdateBoundariesForRange(smallest_candidate, largest_candidate,
-                                     tombstone.seq_,
-                                     cfd->internal_comparator());
-      // The smallest key in a file is used for range tombstone truncation, so
-      // it cannot have a seqnum of 0 (unless the smallest data key in a file
-      // has a seqnum of 0). Otherwise, the truncated tombstone may expose
-      // deleted keys at lower levels.
-      assert(smallest_ikey_seqnum == 0 ||
-             ExtractInternalKeyFooter(meta->smallest.Encode()) !=
-                 PackSequenceAndType(0, kTypeRangeDeletion));
-    }
+  // Add range tombstones
+  auto earliest_snapshot = kMaxSequenceNumber;
+  if (existing_snapshots_.size() > 0) {
+    earliest_snapshot = existing_snapshots_[0];
   }
-  const uint64_t current_entries = sub_compact->builder->NumEntries();
   if (s.ok()) {
-    s = sub_compact->builder->Finish();
+    s = sub_compact->AddRangeTombstones(range_del_agg, range_del_out_stats,
+                                        earliest_snapshot, next_table_min_key);
+  }
+
+  const uint64_t current_entries = sub_compact->GetTableBuilder()->NumEntries();
+  if (s.ok()) {
+    s = sub_compact->GetTableBuilder()->Finish();
   } else {
-    sub_compact->builder->Abandon();
+    sub_compact->GetTableBuilder()->Abandon();
   }
-  IOStatus io_s = sub_compact->builder->io_status();
+
+  IOStatus io_s = sub_compact->GetTableBuilder()->io_status();
   if (s.ok()) {
     s = io_s;
   }
-  const uint64_t current_bytes = sub_compact->builder->FileSize();
+  const uint64_t current_bytes = sub_compact->GetTableBuilder()->FileSize();
   if (s.ok()) {
     meta->fd.file_size = current_bytes;
-    meta->marked_for_compaction = sub_compact->builder->NeedCompact();
+    meta->marked_for_compaction = sub_compact->GetTableBuilder()->NeedCompact();
```

Comment:
meta->oldest_ancester_time appears to be wrong. Perhaps add a TODO?

Reply:
Are you saying the oldest_ancester_time could be improved to be more accurate, right? To do that we need the sequence number to time mapping. Will add todo for that.